apps/subscription.go

```diff
@@ -68,83 +90,84 @@ const (
 // events the app would like to be notified on, and how these notifications
 // should be invoked.
 type Subscription struct {
-	// AppID is used internally by Mattermost. It does not need to be set by app
-	// developers.
-	AppID AppID `json:"app_id"`
+	Event
 
-	// UserID is used internally by Mattermost. It does not need to be set by app
-	// developers.
-	UserID string `json:"user_id"`
+	// Call is the (one-way) call to make upon the event.
+	Call Call `json:"call"`
+}
 
+type Event struct {
 	// Subscription subject. See type Subject godoc (linked) for details.
 	Subject Subject `json:"subject"`
 
 	// ChannelID and TeamID are the subscription scope, as applicable to the subject.
 	ChannelID string `json:"channel_id,omitempty"`
 	TeamID    string `json:"team_id,omitempty"`
-
-	// Call is the (one-way) call to make upon the event.
-	Call Call `json:"call"`
 }
 
 func (sub Subscription) Validate() error {
 	var result error
-	if sub.Subject == "" {
-		result = multierror.Append(result, utils.NewInvalidError("subject most not be empty"))
-	}
-
 	emptyCall := Call{}
 	if sub.Call == emptyCall {
-		result = multierror.Append(result, utils.NewInvalidError("call most not be empty"))
+		result = multierror.Append(result, utils.NewInvalidError("call must not be empty"))
+	}
+	return sub.Event.validate(result)
+}
+
+func (e Event) Validate() error {
+	var result error
+	return e.validate(result)
+}
+
+func (e Event) validate(appendTo error) error {
+	if e.Subject == "" {
+		appendTo = multierror.Append(appendTo, utils.NewInvalidError("subject must not be empty"))
 	}
 
-	switch sub.Subject {
+	switch e.Subject {
+	// Globally scoped, must not contain any extra qualifiers.
 	case SubjectUserCreated,
-		SubjectBotJoinedChannel,
-		SubjectBotLeftChannel,
 		SubjectBotJoinedTeam,
 		SubjectBotLeftTeam /*, SubjectBotMentioned*/ :
-		if sub.TeamID != "" {
-			result = multierror.Append(result, utils.NewInvalidError("teamID must be empty"))
+		if e.TeamID != "" {
+			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is globally scoped; team_id and channel_id must both be empty", e.Subject))
 		}
-		if sub.ChannelID != "" {
-			result = multierror.Append(result, utils.NewInvalidError("channelID must be empty"))
-		}
-
-	case SubjectUserJoinedChannel,
-		SubjectUserLeftChannel /*, SubjectPostCreated */ :
-		if sub.TeamID != "" {
-			result = multierror.Append(result, utils.NewInvalidError("teamID must be empty"))
-		}
-
-		if sub.ChannelID == "" {
-			result = multierror.Append(result, utils.NewInvalidError("channelID must not be empty"))
+		if e.ChannelID != "" {
+			appendTo = multierror.Append(appendTo, utils.NewInvalidError("%s is globally scoped; team_id and channel_id must both be empty", e.Subject))
 		}
 
+	// Team scoped, require TeamID, no ChannelID
 	case SubjectUserJoinedTeam,
 		SubjectUserLeftTeam,
+		SubjectBotJoinedChannel,
+		SubjectBotLeftChannel,
```

Comment:
➕

Reply:
What is the use case for a system-wide subscription? cc @asatkinson

Reply:
I'll assume that the rest of the permissions system handles whether a bot is allowed to join a particular channel.  I think that if the bot is invited into any team it should be able to get a notification that this occurred wouldn't it?

Is there a piece of this I might not be understanding?

Reply:
@asatkinson The "issue" here, is that in order for an App to subscribe to "bot added to channel", the App needs to explicitly create a subscription for a specific team. So the process of receiving a notification for "bot added to channel" is:

- Create a subscription for "bot added to team"
- When the App receives a notification for this event, create another subscription for "bot added to channel" for this specific team

Alternatively, it could be implemented as:

- Create a subscription for "bot added to channel", and have it work for all teams

In the current model, the App needs to monitor team invitations, when it possibly doesn't care about those (it could also miss this notification). It also ends up potentially having noise with the number of subscriptions, when it could just request one centralized "bot added to channel" subscription.

I don't see a use case where the App would want to be notified of being added to a channel in one team, but not some other specific team. If it has such a use case, it can ignore those specific events coming in for those teams it doesn't care about.

We have a relatively small sample size of App developers that we can query for opinions about this, but @dsharma522 has voiced his opinion in this thread implying that it should be implemented as team-agnostic.

Reply:
Concur. Filed https://mattermost.atlassian.net/browse/MM-48793.